### ISSUE TITLE
Is it possible to create a WPF application with .NET 7 that starts in the sys tray?

### ISSUE THREAD

**Issue by ComptonAlvaro** [dotnet/runtime#80088]
I would like to create an application that starts in the sys tray, and I find this question in StackOverflow: [](https://stackoverflow.com/questions/1472633/wpf-application-that-only-has-a-tray-icon)
The answer tells that it is not possible in WPF, but it was requested many times. But this answer is from 2009, so I was wondering if actually it is possibe to do it or not.

Thanks.

**Comment by am11** [comment]
Yes, it is possible. The net7.0-windows app created by `dotnet new wpf` template lacks a few things. This method uses icon as embedded resource:

1. Create project with `dotnet new wpf -n wpf1`.
2. Open wpf1/wpf1.csproj in Visual Studio.
3. Edit wpf1.csproj (as text) and add `<UseWindowsForms>true</UseWindowsForms>` next to `<UseWpf>true</UseWpf>` line (this is for `NotifyIcon`). Save & close the csproj file.
4. Right-click wpf1 in Solution Explorer > Add > New Item > search and select Resource File > name it `ProjectResources.resx` and press Add.
5. Open ProjectResources.resx (in Visual Studio) and click the down arrow next to `Add Resource` > Add Existing File > browse and select the .ico file you want to show in system tray; say `MyIcon.ico`.
6. If you want the app to keep running in the background on close with its icon in system tray: In App.xaml, add `ShutdownMode="OnExplicitShutdown"` attribute in `Application` tag.
7. In App.xaml.cs, apply this patch:
    ```diff
    ...
    + using NotifyIcon = System.Windows.Forms.NotifyIcon;

    namespace wpf1
    {
    ...
        public partial class App : Application
        {
    +        protected override void OnStartup(StartupEventArgs e)
    +        {
    +            NotifyIcon icon = new()
    +            {
    +                icon.Icon = ProjectResources.MyIcon,
    +                icon.Visible = true
    +            };
    +            icon.Click += new EventHandler(TrayIconOnClick);
    +
    +            base.OnStartup(e);
    +
    +            void TrayIconOnClick(object? sender, EventArgs e) =>
    +                MessageBox.Show("Tray icon clicked!");
    +       }
        }
    ...
    ```
8. Save All, Build All and Run the app from VS or `dotnet run`. You'll see your icon in system tray.

**Comment by ComptonAlvaro** [comment]
thanks for the code example. But I have a problem in the step 5 and 7.

I have created the ProjectResources.resx file and I have added a ico to them. It is recognize in the code in 7, but when I try to set the icon.Icon wht ProjectResources.MyIcon, I get an error that tells: cannont initialize type NotifyIcon with a collection initializer that doesn't implement System.Collections.IEnumerable.

**Comment by am11** [comment]
If you don't want the MainWindow to appear on app startup, you can simply remove `StartupUri="MainWindow.xaml"` from App.xaml file.

**Comment by ViktorHofer** [comment]
Moving to the dotnet/wpf repository.
